src/kvp.rs

```diff
@@ -0,0 +1,684 @@
+fn generate_event_key(
+    event_level: &str,
+    event_name: &str,
+    span_id: &str,
+) -> String {
+    format!(
+        "{}|{}|{}|{}",
+        EVENT_PREFIX, event_level, event_name, span_id
+    )
+}
+
+/// Encodes a key-value pair (KVP) into one or more byte slices. If the value
+/// exceeds the allowed size, it is split into multiple slices for encoding.
+/// This is used for logging events to a KVP file.
+///
+/// # Arguments
+/// * `key` - The key as a string slice.
+/// * `value` - The value associated with the key.
+fn encode_kvp_item(key: &str, value: &str) -> Vec<Vec<u8>> {
+    let key_bytes = key.as_bytes();
+    let value_bytes = value.as_bytes();
+
+    let key_len = key_bytes.len().min(HV_KVP_EXCHANGE_MAX_KEY_SIZE);
+    let mut key_buf = vec![0u8; HV_KVP_EXCHANGE_MAX_KEY_SIZE];
+    key_buf[..key_len].copy_from_slice(&key_bytes[..key_len]);
+
+    let mut kvp_slices = Vec::new();
+
+    for chunk in value_bytes.chunks(HV_KVP_AZURE_MAX_VALUE_SIZE) {
+        let mut value_buf = vec![0u8; HV_KVP_EXCHANGE_MAX_VALUE_SIZE];
+        value_buf[..chunk.len()].copy_from_slice(chunk);
+
+        kvp_slices.push(encode_kvp_slice(key_buf.clone(), value_buf));
+    }
+
+    kvp_slices
+}
+
+/// Combines the key and value of a KVP into a single byte slice, ensuring
+/// proper formatting for consumption by hv_kvp_daemon service,
+/// which typically reads from /var/lib/hyperv/.kvp_pool_1.
+fn encode_kvp_slice(key: Vec<u8>, value: Vec<u8>) -> Vec<u8> {
+    let mut buffer = Vec::with_capacity(
+        HV_KVP_EXCHANGE_MAX_KEY_SIZE + HV_KVP_EXCHANGE_MAX_VALUE_SIZE,
+    );
+    buffer.extend_from_slice(&key);
+    buffer.extend_from_slice(&value);
+    buffer
+}
+
+/// Decodes a KVP byte slice into its corresponding key and value strings.
+/// This is useful for inspecting or logging raw KVP data.
+#[cfg(test)]
+pub fn decode_kvp_item(
+    record_data: &[u8],
+) -> Result<(String, String), &'static str> {
+    let record_data_len = record_data.len();
+    let expected_len =
+        HV_KVP_EXCHANGE_MAX_KEY_SIZE + HV_KVP_EXCHANGE_MAX_VALUE_SIZE;
+
+    if record_data_len != expected_len {
+        return Err("record_data len not correct.");
+    }
+
+    let key = String::from_utf8(
+        record_data[0..HV_KVP_EXCHANGE_MAX_KEY_SIZE].to_vec(),
+    )
+    .unwrap_or_else(|_| String::new())
+    .trim_end_matches('\x00')
+    .to_string();
+
+    let value = String::from_utf8(
+        record_data[HV_KVP_EXCHANGE_MAX_KEY_SIZE..]
+            .iter()
+            .take(HV_KVP_AZURE_MAX_VALUE_SIZE)
+            .cloned()
+            .collect(),
+    )
+    .unwrap_or_else(|_| String::new())
+    .trim_end_matches('\x00')
+    .to_string();
+
+    Ok((key, value))
+}
+
+/// Truncates the guest pool KVP file if it contains stale data (i.e., data
+/// older than the system's boot time). Logs whether the file was truncated
+/// or no action was needed.
+fn truncate_guest_pool_file(kvp_file: &Path) -> Result<(), Error> {
```

Comment:
Optional, but since this lives in the `azure-init` crate you can use the `anyhow::Error` type to avoid having to do the `map_err` stuff.